src/causalprog/graph/graph.py

```diff
@@ -0,0 +1,45 @@
+"""Graph storage."""
+
+import networkx as nx
+
+from .node import Node
+
+
+class Graph:
+    """A directed acyclic graph that represents a causality tree."""
+
+    def __init__(self, graph: nx.Graph, label: str) -> None:
+        """Initialise a graph from a NetworkX graph."""
+        for node in graph.nodes:
+            if not isinstance(node, Node):
+                msg = f"Invalid node: {node}"
+                raise TypeError(msg)
+
+        self._label = label
+
+        self._graph = graph
+        self._nodes = list(graph.nodes())
+        self._depth_first_nodes = list(nx.algorithms.dfs_postorder_nodes(graph))
+
+        self._predecessors = nx.algorithms.dfs_predecessors(graph)
+        self._successors = nx.algorithms.dfs_successors(graph)
```

Comment:
Do we want to take a copy of `graph` here, just in case the user decides to edit the `nx.Graph` object in their script / analysis later and doesn't appreciate the consequences of doing so?

Setting `self._predecessors` (and the others attributes) implies that we want to be able to treat `self._graph` as immutable. But because Python is Python, there is still scope for errors due to changing the underlying objects.

There might also be a case for making `_predecessors` (etc) into `properties`, but I'm guessing they'd rather be attributes so we can set them once and don't have to re-run the `dfs` algorithm over and over again.